gk/main.c

```diff
@@ -1175,6 +1245,384 @@ gk_conf_put(struct gk_config *gk_conf)
+ * and holds the lock of that FIB entry.
+ */
+static struct gk_fib *
+add_prefix_fib(struct ipaddr *ip_addr,
+	int prefix_len, struct gk_config *gk_conf)
+{
+	int ret;
+	int fib_id = -1;
+	struct ipv4_lpm_route ipv4_route;
+	struct ipv6_lpm_route ipv6_route;
+	struct gk_fib *new_fib = NULL;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	/* Find an empty fib entry for the IP address. */
+	if (ip_addr->proto == ETHER_TYPE_IPv4) {
+		if (!ipv4_if_configured(&gk_conf->net->back))
+			goto out;
+
+		fib_id = get_empty_fib_id(ltbl->fib_tbl);
+		if (fib_id < 0)
+			goto out;
+
+		new_fib = &ltbl->fib_tbl[fib_id];
+
+		/*
+		 * Add the fib entry for the IPv4 address
+		 * to the IPv4 LPM table.
+		 */
+		ipv4_route.ip = ntohl(ip_addr->ip.v4.s_addr);
+		ipv4_route.depth = prefix_len;
+		ipv4_route.nexthop = fib_id;
+		ret = gk_lpm_add_ipv4_routes(
+			ltbl->lpm, &ipv4_route, 1, ltbl);
+		if (ret < 0)
+			goto out;
+	} else if (ip_addr->proto == ETHER_TYPE_IPv6) {
+		if (!ipv6_if_configured(&gk_conf->net->back))
+			goto out;
+
+		fib_id = get_empty_fib_id(ltbl->fib_tbl6);
+		if (fib_id < 0)
+			goto out;
+
+		new_fib = &ltbl->fib_tbl6[fib_id];
+
+		/*
+		 * Add the fib entry for the IPv6 address 
+		 * to the IPv6 LPM table.
+		 */
+		rte_memcpy(ipv6_route.ip,
+			ip_addr->ip.v6.s6_addr, sizeof(ipv6_route.ip));
+		ipv6_route.depth = prefix_len;
+		ipv6_route.nexthop = fib_id;
+		ret = gk_lpm_add_ipv6_routes(
+			ltbl->lpm6, &ipv6_route, 1, ltbl);
+		if (ret < 0)
+			goto out;
+	} else
+		goto out;
+
+	new_fib->ref_cnt = 1;
+
+out:
+	return new_fib;
+}
+
+static int
+setup_neighbor_tbl(unsigned int socket_id, int identifier,
+	int ip_ver, int ht_size, struct neighbor_hash_table *neigh)
+{
+	int  ret;
+	char ht_name[64];
+	int key_len = ip_ver == ETHER_TYPE_IPv4 ?
+		sizeof(struct in_addr) : sizeof(struct in6_addr);
+
+	struct rte_hash_parameters neigh_hash_params = {
+		.entries = ht_size,
+		.key_len = key_len,
+		.hash_func = DEFAULT_HASH_FUNC,
+		.hash_func_init_val = 0,
+	};
+
+	ret = snprintf(ht_name, sizeof(ht_name),
+		"neighbor_hash_%u", identifier);
+	RTE_VERIFY(ret > 0 && ret < (int)sizeof(ht_name));
+
+	neigh->tbl_size = ht_size;
+
+	/* Setup the neighbor hash table. */
+	neigh_hash_params.name = ht_name;
+	neigh_hash_params.socket_id = socket_id;
+	neigh->hash_table = rte_hash_create(&neigh_hash_params);
+	if (neigh->hash_table == NULL) {
+		RTE_LOG(ERR, HASH,
+			"The GK block cannot create hash table for neighbor FIB!\n");
+
+		ret = -1;
+		goto out;
+	}
+
+	/* Setup the Ethernet header cache table. */
+	neigh->cache_tbl = rte_calloc(NULL,
+		ht_size, sizeof(struct ether_cache), 0);
+	if (neigh->cache_tbl == NULL) {
+		RTE_LOG(ERR, MALLOC,
+			"The GK block cannot create Ethernet header cache table\n");
+
+		ret = -1;
+		goto neigh_hash;
+	}
+
+	ret = 0;
+	goto out;
+
+neigh_hash:
+	rte_hash_free(neigh->hash_table);
+	neigh->hash_table = NULL;
+out:
+	return ret;
+}
+
+static int
+init_fib_tbl(struct gk_config *gk_conf)
```

Comment:
This function is so long that is hard to make sure that it's correct. Moreover, it is doing mostly trivial stuff. I suggest to break it up.